File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionMatchAll.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.util.ArrayList;
import java.util.Map;
import java.util.regex.Pattern;

/*
 * For tests
 * @see TransformerFunctionMatchTest
 */
public class TransformerFunctionMatchAll extends TransformerFunction {
    public TransformerFunctionMatchAll() {
        super(FunctionDescription.of(
            Map.of(
                "pattern", ArgumentType.of(ArgType.String).position(0),
                "group", ArgumentType.of(ArgType.Integer).position(1).defaultInteger(0)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var str = context.getString(null);
        if (str == null) {
            return null;
        }
        var patternString = context.getString("pattern");
        if (patternString == null) {
            return null;
        }
        var matcher = Pattern.compile(patternString).matcher(str);
        var group = context.getInteger("group");
        var allMatches = new ArrayList<>();
        while (matcher.find()) {
            allMatches.add(matcher.group(group));
        }
        return allMatches.size() == 0 ? null : allMatches ;
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionMatchAll.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";

class TransformerFunctionMatchAll extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        pattern: { type: ArgType.String, position: 0 },
        group: { type: ArgType.Integer, position: 1, defaultInteger: 0 },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const str = await context.getString(null);
    if (str == null) {
      return null;
    }
    const patternString = await context.getString("pattern");
    if (patternString == null) {
      return null;
    }
    const matcher = new RegExp(patternString, "g");

    const allMatches: string[] = [];
    const group = await context.getInteger("group");
    for (const match of str.matchAll(matcher)) {
      allMatches.push(match[group ?? 0]);
    }
    return allMatches.length == 0 ? null : allMatches;
  }
}

export default TransformerFunctionMatchAll;
